### ISSUE TITLE
add_child seems very slow under pure java implementation

### ISSUE THREAD

**Issue by ggodet-bar** [sparklemotion/nokogiri#692]
Hi,

I'm running the following script

```ruby
require 'nokogiri'
require 'benchmark'

puts Nokogiri::VERSION

doc = Nokogiri::XML::Document.new
root = Nokogiri::XML::Node.new('root', doc)

Benchmark.bm { |x|
  x.report {
    10000.times do |i|
      root << Nokogiri::XML::Node.new('randomNode', doc)
    end
  }
}
```

Running under 1.9.3-p125 returns the following

```
1.5.2
       user     system      total        real
   0.030000   0.000000   0.030000 (  0.027721)
```

... while running on JRuby (1.6.7, 1.7.0.preview1 and head) return uniformly

```
1.5.2
       user     system      total        real
  21.880000   0.100000  21.980000 ( 21.259000)
```

When running JRuby's profiler, it turns out that 95% of these twenty-something seconds is spent in the add_child method.

This might be an unfortunate but natural side-effect of using the Xerces library, but I haven't found any explicit documentation about this.

**Comment by flavorjones** [comment]
@yokolet @jvshahid This still appears to be a poorly performing method.

```
# ruby 2.3.0
1.6.7.2
       user     system      total        real
   0.020000   0.000000   0.020000 (  0.020321)
```

```
# jruby 9.0.4.0
1.6.7.2
       user     system      total        real
  84.110000   0.090000  84.200000 ( 83.353718)
```

Any thoughts?

**Comment by jvshahid** [comment]
I'm not familiar with this issue, but i have the benchmark on my dev machine and will investigate it.

**Comment by jvshahid** [comment]
Verified that the call to `relink_namepsace` is taking long time. I also attached an image of a profiling session. Removing out the `relink_namespace` call (as pointed out previously) improves the  performance dramatically as shown below:

```
       user     system      total        real
   0.230000   0.000000   0.230000 (  0.190405)
```

![snapshot-1455818721109-cpu-calltree](https://cloud.githubusercontent.com/assets/297621/13153252/1e6133f8-d641-11e5-81b8-d43548bb4575.png)

**Comment by jvshahid** [comment]
I finally got time to figure out what's going on. It seems like the JRuby implementation had a bug that caused it to run in O(n<sup>2</sup>). The `relink_namespace` method was called on the `root` node every time a new node is added as a child of `root`. I modified the code locally and it seems to have fixed the slowness issue. I will take the opportunity to clean up the `relink_namespace` code and document it. i also need to figure out a way to test this behavior in order to prevent regressions.
